unblob/file_utils.py

```diff
@@ -105,44 +119,39 @@ def find_first(
-    while True:
-        current_position = file.tell()
-
-        # Prepend the padding from the last chunk, to make sure that we find the pattern,
-        # even if it straddles the chunk boundary.
-        data = file.read(chunk_size)
-        if data == b"":
-            # We've reached the end of the stream.
-            file.seek(initial_position)
-            return -1
-
-        marker = data.find(pattern)
-        if marker != -1:
-            found_pos = current_position + marker
-            # We want to seek past the found position to the next byte,
-            # so we can call find_first again without extra seek
-            # This might seek past the actual end of the file
-            file.seek(found_pos + len(pattern))
-            return found_pos
-
-        if len(data) < len(pattern):
-            # The length that we read from the file is the same length or less than as the pattern
-            # we're looking for, and we didn't find the pattern in there. If we don't return -1
-            # here, we'll end up in an infinite loop.
-            file.seek(initial_position)
-            return -1
-
-        file.seek(-compensation, os.SEEK_CUR)
-        bytes_searched += chunk_size - compensation
-
-
-def iterate_patterns(file: io.BufferedIOBase, pattern: bytes, chunk_size: int = 0x1000):
-    """Iterate on the file searching for pattern until all occurences has been found."""
-    while True:
-        match = find_first(file, pattern, chunk_size)
-        if match == -1:
-            return
-        yield match
+    try:
+        while True:
+            current_position = file.tell()
+
+            # Prepend the padding from the last chunk, to make sure that we find the pattern,
+            # even if it straddles the chunk boundary.
+            data = file.read(chunk_size)
+            if data == b"":
+                # We've reached the end of the stream.
+                return
+
+            if len(data) < len(pattern):
+                # The length that we read from the file is the same
+                # length or less than as the pattern we're looking
+                # for, and we didn't find the pattern in there.
+                return
+
+            marker = data.find(pattern)
+            while marker != -1:
+                found_pos = current_position + marker
+                # We want to seek past the found position to the next byte,
+                # so we can call find_first again without extra seek
+                # This might seek past the actual end of the file
+                file.seek(found_pos + len(pattern))
+                yield found_pos
+                # re-set position in-case it gets changed during iteration
+                file.seek(found_pos + len(pattern))
+                marker = data.find(pattern, marker + len(pattern))
+
+            file.seek(-compensation, os.SEEK_CUR)
+            bytes_searched += chunk_size - compensation
+    finally:
+        file.seek(initial_position)
```

Comment:
I like the idea to always reset the file pointer. A bunch of tests fail now because of this. What do you think, should I adjust the tests or restore the old behavior?

Reply:
I like that idea too as it will make the behavior generic between the different cases (no pattern found, one pattern found, multiple pattern found). So I'd go for adjusting tests.

If I understand correctly, the iterator consumer must read all values from it for the pointer to be set back, right ?

Reply:
Or discard the iterator, like in the case of `find_first` Another possible way to go is to reset the pointer to the beginning at each yield, so it is the consumer's responsibility to seek.

Reply:
I have gone forward with the latter approach, only UBI depended on the seeking side-effect anyway.